File: LeetCode_Algorithm/131(Palindrome Partitioning)/131.py
'''
LeetCode 131 : Palindrome Partitioning
주어진 스트링에서 가능한 서브스트링이 회문인 모든 가짓수를 리턴하라.

# sol1 688ms 76% 28.8MB 96%
O(n*2^n) O(n*2^n)
"aab"에서 가질 수 있는 모든 서브스트링은 
a aa aab이다
이 과정에서 서브 스트링이 회문이라면 그 서브스트링만 재귀로 넘긴다.


'''
class Solution:
    def partition(self, s: str) -> list[list[str]]:
        self.ans = []
        self.solve(s, [])
        return self.ans
    def solve(self, s, temp):
        if not s:
            self.ans.append(temp[:])
        for i in range(1, len(s)+1):
            if self.isPalindrome(s[:i]):
                temp.append(s[:i])
                self.solve(s[i:], temp)
                temp.pop()
            
    def isPalindrome(self, s):
        for i in range(len(s)//2):
            if s[i] != s[len(s)-i-1]:
                return False
        return True

print(Solution().partition("aab"))

File: LeetCode_Algorithm/131(Palindrome Partitioning)/131.js
var isPalindrome = (s) => {
  for (let i=0; i < parseInt(s.length / 2); i++) {
    if (s[i] != s[s.length-1-i])
      return false
  }
  return true
}

var solve = (s, temp) => {
  if (!s)
    ans.push(temp.slice(0, ));
  for (let i=1;i < s.length+1; i++) {
    if (isPalindrome(s.slice(0,i))) {
      temp.push(s.slice(0, i))
      solve(s.slice(i,), temp)
      temp.pop()
    }
  }
}

var partition = function(s) {
  // sol1 250ms 89% 63.3MB 75%
  ans = []
  solve(s, [])
  return ans
};
console.log(partition("aab"))